otherlibs/systhreads/st_win32.h

```diff
@@ -38,18 +38,46 @@ typedef DWORD st_retcode;
 
 #define SIGPREEMPTION SIGTERM
 
+/* Unique thread identifiers and atomic operations over them */
+#ifdef ARCH_SIXTYFOUR
+typedef LONG64 st_tid;
+#define Tid_Atomic_Exchange InterlockedExchange64
+#define Tid_Atomic_Compare_Exchange InterlockedCompareExchange64
+#else
+typedef LONG st_tid;
+#define Tid_Atomic_Exchange InterlockedExchange
+#define Tid_Atomic_Compare_Exchange InterlockedCompareExchange
+#endif
+
+/* Thread-local storage associating a thread identifier to every thread. */
+static DWORD st_thread_tid_key;
+
 /* Thread-local storage associating a Win32 event to every thread. */
 static DWORD st_thread_sem_key;
 
+/* Next thread identifier -- 0 is reserved */
+static st_tid st_next_thread_id = 1;
+
+static st_tid st_fresh_thread_id (void)
+{
+  st_tid res = st_next_thread_id ++;
+  /* Guard against overflow */
+  if (res == 0) res = st_next_thread_id ++;
```

Comment:
I do not understand this piece:
* This is assuming a signed overflow, right?
* You seem to assume that one cannot get there for realistic programs even on 32-bit, right? In that case is it not better to fail properly with an exception?

Reply:
Yes, no, yes.  But the Win32 code you reviewed was broken elsewhere, so I just pushed an alternate approach that uses the OCaml thread identifier instead.  Sorry for the back and forth.